src/lexer.coffee

```diff
@@ -555,18 +555,40 @@ exports.Lexer = class Lexer
         prev.spaced or
         prev[0] not in COMPARABLE_LEFT_SIDE
       )
-      [input, id, colon] = match
-      origin = @token 'CSX_TAG', id, offset: 1, length: id.length
-      @token 'CALL_START', '('
-      @token '[', '['
-      @ends.push tag: '/>', origin: origin, name: id
+      [input, id] = match
+      fullId = id
+      if '.' in id
+        [id, properties...] = id.split '.'
+      else
+        properties = []
+      tagToken = @token 'CSX_TAG', id,
+        length: id.length + 1
+        data:
+          openingBracketToken: @makeToken '<', '<'
```

Comment:
> Why is `<` a “sub-token” rather than a delimiter like strings’ `"`?

Strings we tokenize as strings and parse as strings. What JSX exposes as tokens for parsing doesn't resemble the actual source code - we put in fake tokens so that JSX gets parsed as nonexistent calls, arrays etc

eg `<div a="b" />` gets "transformed" so that to the parser it looks like `div([{a: 'b'}])`

But in the process we'd be losing track of all the actual original source code location data

So we need to preserve the original JSX location data as we're doing the transformation

> Or if we need to present it as a node in the AST, why not generate it in the node classes rather instead of producing this unparsed psuedo-token here?

By that time, we'd have lost the original location data. It's not about generating AST nodes using the sub-tokens, it's about generating accurate AST location data

Reply:
>  It’s not about generating AST nodes using the sub-tokens, it’s about generating accurate AST location data

What about the location data for JSX code in the normal compiler, not for AST output? Are we losing a big chunk of the source maps location data right now, that we should be preserving? Should we perhaps make these lexer changes on a new branch off of `master`, to improve JSX location data in the normal compiler?

Reply:
I don't think I'm in a position to (easily) assess how badly source maps suffer due to the way we're handling JSX, or how big of a deal that is. But my rough understanding is that if source maps are not handling JSX accurately, we'd probably have to fully parse JSX in order to have accurate JSX location data propagated to source maps

Given that most of the lexer changes here are about exposing the "sub-tokens" (which wouldn't affect source maps) I don't think that it'd be worth trying to target this specifically for that purpose